### DISCUSSION TITLE
Get row by index in a Dask DataFrame

### DISCUSSION THREAD

**Discussion by danilopeixoto** [dask/dask#10935] [Q&A]
What is the appropriate method for obtaining a specific row by index in a Dask DataFrame?

**Source**

```python
import dask.dataframe as dd


class Dataset:
    def __init__(self, path: str):
        self.df = dd.read_csv(path)

    def __len__(self) -> int:
        return len(self.df)

    def __getitem__(self, index: int) -> list:
        return self.df.loc[index].compute().values[0].tolist()


if __name__ == '__main__':
    dataset = Dataset('wine_dataset/large.csv')

    first_index = 0
    last_index = len(dataset) - 1

    print(f'{first_index=}')
    print(f'{last_index=}')

    print(f'{dataset[first_index]=}')
    print(f'{dataset[last_index]=}')
```

**Output**

```txt
first_index=0
last_index=155435343
dataset[first_index]=[6.4, 0.28, 0.29, 1.6, 0.052, 34.0, 127.0, 0.9929, 3.48, 0.56, 10.5, 7.0]
Traceback (most recent call last):
  File "dataset.py", line 25, in <module>
    print(f'{dataset[last_index]=}')
  File "dataset.py", line 12, in __getitem__
    return self.df.loc[index].compute().values[0].tolist()
IndexError: index 0 is out of bounds for axis 0 with size 0
```

**Comment by TayfurYldz** [comment]
The integer passed to `.loc` is an **index label**, not a global row position. That is why `len(df) - 1` is not a valid way to address the last row here.

This matters especially for `dd.read_csv`: a Dask DataFrame is partitioned, and the default pandas-style indexes in those partitions do not imply a globally materialized `RangeIndex(0, len(df))`. `len(df)` computes a total row count, but it does not create a label with value `len(df)-1`.

Dask also deliberately does not provide pandas-style positional row `.iloc[n]`. On current `main`, `ILocIndexer` rejects row indexers and says that `DataFrame.iloc` only supports selecting columns:

```python
df.iloc[:, column_indexer]   # supported
df.iloc[n]                   # not a supported Dask operation
```

If you mean a **label**, establish an index with the semantics you need and use `.loc[label]`. If you mean “the last row”, use the operation that expresses that directly:

```python
row = df.tail(1).iloc[0].tolist()
```

(`tail(1)` computes a small pandas result, so pandas `.iloc[0]` is fine after that.)

For arbitrary positional random access — implementing a PyTorch-style `Dataset.__getitem__(n)` over a CSV — a Dask DataFrame is generally the wrong abstraction. To locate global row `n`, something first has to know the row counts/offsets of the partitions; doing that repeatedly defeats much of the point of lazy partitioned execution. If random row access is a real workload requirement, create/persist an explicit stable row-id/index during preprocessing or use a storage format/index designed for random access.

So the immediate bug in the example is `.loc[index]` being treated as positional indexing; it is label-based indexing.
